climada/hazard/tc_tracks.py

```diff
@@ -322,6 +415,68 @@ def subset(self, filterdict):
 
         return out
 
+    def subset_by_basin(self):
+        """Subset all tropical cyclones tracks by basin.
+
+        This function iterates through the tropical cyclones in the dataset and assigns each cyclone
+        to a basin based on its geographical location. It checks whether the cyclone's position
+        (latitude and longitude) lies within the boundaries of any of the predefined basins and
+        then groups the cyclones into separate categories for each basin. The resulting dictionary
+        maps each basin's name to a list of tropical cyclones that fall within it.
+
+        Parameters
+        ----------
+        self : TCTtracks object
+            The object instance containing the tropical cyclone data (`self.data`) to be processed.
+
+        Returns
+        -------
+        dict_tc_basins : dict
+            A dictionary where the keys are basin names (e.g., "NA", "EP", "WP", etc.) and the
+            values are instances of the `TCTracks` class containing the tropical cyclones that
+            belong to each basin.
+
+        Example:
+        --------
+        >>> tc = TCTracks.from_ibtracks("")
+        >>> tc_basins = tc.split_by_basin()
+        >>> tc_basins["NA"] # to access tracks in the North Atlantic
+
+        """
+
+        # Initialize a defaultdict to store lists for each basin
+        basins_dict = defaultdict(list)
+        tracks_outside_basin: list = []
+        # Iterate over each tropical cyclone
+        for track in self.data:
+            lat, lon = track.lat.values[0], track.lon.values[0]
+            origin_point = Point(lon, lat)
+            point_in_basin = False
+
+            # Find the basin that contains the point
+            for basin in Basin:
+                if basin.value.contains(origin_point):
+                    basins_dict[basin.name].append(track)
+                    point_in_basin = True
+                    break
+
+            if not point_in_basin:
+                tracks_outside_basin.append(track.id_no)
```

Comment:
@NicolasColombi as to 1. and 2.
I would first define a method that assigns basins to a track:
```python
BASINS_GDF = gpd.GeoDataFrame(
    {'basin': b, 'geometry': b.value} for b in Basin
)

def get_basins(track):  # this is the method I had in mind for 1. and I'd guess it could be a performance boost
    track_coordinates = GeoDataFrame(
        geometry=gpd.points_from_xy(track.lon, track.lat)
    )
    return track_coordinates.sjoin(BASINS_GDF , how='left', predicate='within').basin
```

Reply:
Having that I'd write the track loop like this:
```python
for track in self.data:
    touched = get_basins(track).dropna().drop_duplicates()
    if touched.size:
        for basin in touched:
            basins_dict[basin].append(track)
    else:
        tracks_outside_basin.append(track)
```

Reply:
I have no strong opinion, but I would say the second makes more sense as a default. Ideally, just allow for both.

Reply:
Ok! I implemented Emanuel version and allowed the flexibility of choosing only the origin basin or not.